<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ MIPanalyzer</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Function reference" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MIPanalyzer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/test1.html">Vignette Title</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="MIPanalyzerbi2vcfR.html">MIPanalyzerbi2vcfR()</a></code> </p>
        </td>
        <td><p>Convert a MIPanalyzer Biallelic Object to a vcfR object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Pf_chrom_lengths.html">Pf_chrom_lengths()</a></code> </p>
        </td>
        <td><p>Get dataframe of P.falciparum chromosome lengths</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assignGTfrombiWSNRAF.html">assignGTfrombiWSNRAF()</a></code> </p>
        </td>
        <td><p>Assign the GT calls from the NONREFERENT WSAF</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="explore_filter_coverage_loci.html">explore_filter_coverage_loci()</a></code> </p>
        </td>
        <td><p>Explore locus coverage prior to filtering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="explore_filter_coverage_samples.html">explore_filter_coverage_samples()</a></code> </p>
        </td>
        <td><p>Explore sample coverage prior to filtering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_coverage_loci.html">filter_coverage_loci()</a></code> </p>
        </td>
        <td><p>Filter loci based on coverage</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_coverage_samples.html">filter_coverage_samples()</a></code> </p>
        </td>
        <td><p>Filter samples based on coverage</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_loci.html">filter_loci()</a></code> </p>
        </td>
        <td><p>Filter out some loci</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_overcounts.html">filter_overcounts()</a></code> </p>
        </td>
        <td><p>Filter out over-counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_samples.html">filter_samples()</a></code> </p>
        </td>
        <td><p>Filter out some samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_IBS_distance.html">get_IBS_distance()</a></code> </p>
        </td>
        <td><p>Get identity by state (IBS) distance</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_genomic_distance.html">get_genomic_distance()</a></code> </p>
        </td>
        <td><p>Get genomic distance between samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_spatial_distance.html">get_spatial_distance()</a></code> </p>
        </td>
        <td><p>Get great circle distance between spatial points</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_wsaf.html">get_wsaf()</a></code> </p>
        </td>
        <td><p>Get within-sample allele frequencies</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="inbreeding_mle.html">inbreeding_mle()</a></code> </p>
        </td>
        <td><p>Estimate pairwise inbreeding coefficient F by maximum likelihood</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="is.mipanalyzer_biallelic.html">is.mipanalyzer_biallelic()</a></code> </p>
        </td>
        <td><p>Determine if object is of class mipanalyzer_biallelic</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lonlat_to_bearing.html">lonlat_to_bearing()</a></code> </p>
        </td>
        <td><p>Calculate great circle distance and bearing between coordinates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mipanalyzer_file.html">mipanalyzer_file()</a></code> </p>
        </td>
        <td><p>Load system file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pca_wsaf.html">pca_wsaf()</a></code> </p>
        </td>
        <td><p>PCA of within-sample allele frequencies</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pcoa_genomic_distance.html">pcoa_genomic_distance()</a></code> </p>
        </td>
        <td><p>PCoA of genomic distances between samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_coverage.html">plot_coverage()</a></code> </p>
        </td>
        <td><p>Plot coverage matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_map.html">plot_map()</a></code> </p>
        </td>
        <td><p>Produce ggplot map</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_pca.html">plot_pca()</a></code> </p>
        </td>
        <td><p>Plot PCA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_pca_contribution.html">plot_pca_contribution()</a></code> </p>
        </td>
        <td><p>Plot PCA contribution of each variable</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_pca_variance.html">plot_pca_variance()</a></code> </p>
        </td>
        <td><p>Plot variance explained by PCA components</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_pcoa.html">plot_pcoa()</a></code> </p>
        </td>
        <td><p>Plot PCoA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="print.mipanalyzer_biallelic.html">print(<i>&lt;mipanalyzer_biallelic&gt;</i>)</a></code> </p>
        </td>
        <td><p>Custom print function for class mipanalyzer_biallelic</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="print_full.html">print_full()</a></code> </p>
        </td>
        <td><p>Ordinary print function for class mipanalyzer_biallelic</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rbetabinom.html">rbetabinom()</a></code> </p>
        </td>
        <td><p>Draw from Beta-binomial distribution</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rdirichlet.html">rdirichlet()</a></code> </p>
        </td>
        <td><p>Draw from Dirichlet distribution</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sim_biallelic.html">sim_biallelic()</a></code> </p>
        </td>
        <td><p>Simulate biallelic data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summary.mipanalyzer_biallelic.html">summary(<i>&lt;mipanalyzer_biallelic&gt;</i>)</a></code> </p>
        </td>
        <td><p>Print summary for class mipanalyzer_biallelic</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vcf2mipanalyzer_biallelic.html">vcf2mipanalyzer_biallelic()</a></code> </p>
        </td>
        <td><p>Convert vcf to biallelic mipanalyzer data class</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Bob Verity.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

